src/sources/datadog/logs.rs

```diff
@@ -0,0 +1,291 @@
+use crate::{
+    config::{
+        log_schema, DataType, GenerateConfig, GlobalOptions, Resource, SourceConfig,
+        SourceDescription,
+    },
+    event::Event,
+    shutdown::ShutdownSignal,
+    sources::{
+        self,
+        http::{decode_body, Encoding},
+        util::{ErrorMessage, HttpSource, HttpSourceAuthConfig},
+    },
+    tls::TlsConfig,
+    Pipeline,
+};
+use bytes::Bytes;
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, net::SocketAddr};
+
+use warp::http::HeaderMap;
+
+lazy_static! {
+    static ref API_KEY_MATCHER: Regex =
+        Regex::new(r"^/v1/input/(?P<api_key>[[:alnum:]]{32})/??").unwrap();
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct DatadogLogsConfig {
+    address: SocketAddr,
+    tls: Option<TlsConfig>,
+    auth: Option<HttpSourceAuthConfig>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<DatadogLogsConfig>("datadog_logs")
+}
+
+impl GenerateConfig for DatadogLogsConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            address: "0.0.0.0:8080".parse().unwrap(),
+            tls: None,
+            auth: None,
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "datadog_logs")]
+impl SourceConfig for DatadogLogsConfig {
+    async fn build(
+        &self,
+        _: &str,
+        _: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<sources::Source> {
+        let source = DatadogLogsSource {};
+        // We accept /v1/input & /v1/input/<API_KEY>
+        source.run(
+            self.address,
+            "/v1/input",
+            false,
+            &self.tls,
+            &self.auth,
+            out,
+            shutdown,
+        )
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "datadog_logs"
+    }
+
+    fn resources(&self) -> Vec<Resource> {
+        vec![Resource::tcp(self.address)]
+    }
+}
+
+#[derive(Clone, Default)]
+struct DatadogLogsSource {}
+
+impl HttpSource for DatadogLogsSource {
+    fn build_event(
+        &self,
+        body: Bytes,
+        header_map: HeaderMap,
+        _query_parameters: HashMap<String, String>,
+        request_path: &str,
+    ) -> Result<Vec<Event>, ErrorMessage> {
+        if body.is_empty() {
+            // The datadog agent may sent empty payload as keep alive
+            debug!(message = "Empty payload ignored.");
+            return Ok(Vec::new());
+        }
+
+        let api_key = extract_api_key(&header_map, request_path).map(|mut k| {
+            if k.len() > 5 {
+                k.replace_range(0..k.len() - 5, "***************************");
```

Comment:
Do we want to mask the key? It seems like might need it later if, for example, we are passing through the `datadog_logs` sink.

Reply:
That was supposed to be an interim measure to avoid leaking key (as we don't really use it anywhere) but after rethinking it, I think it should be a note in the documentation, suggesting to use `remove_fields` to strip the key from received event.